examples/nlp/data_parallel_training_with_keras_nlp.py

```diff
@@ -0,0 +1,202 @@
+industry workflows.
+- On a cluster of many machines, each hosting one or multiple GPUs (multi-worker
+distributed training). This is a good setup for large-scale industry workflows, e.g.
+training high-resolution image classification models on tens of millions of images using
+20-100 GPUs.
+
+## Setup
+
+The tutorial relies on KerasNLP 0.5.2. Additionally, we need
+at least TensorFlow 2.11 in order to use AdamW with mixed precision.
+"""
+
+"""shell
+pip install keras-nlp==0.5.2 -q
+pip install -U tensorflow -q
+"""
+
+"""
+## Imports
+"""
+
+import os
+import tensorflow as tf
+from tensorflow import keras
+import keras_nlp
+
+"""
+A simple thing we can do right off the bat is to enable
+[mixed precision](https://keras.io/api/mixed_precision/), which will speed up training by
+running most of our computations with 16 bit (instead of 32 bit) floating point numbers.
+"""
+
+policy = keras.mixed_precision.Policy("mixed_float16")
+keras.mixed_precision.set_global_policy(policy)
+
+
+PRETRAINING_BATCH_SIZE = 128
+EPOCHS = 5
+
+"""
+First, we need to download and preprocess the wikitext-2 dataset. This dataset will be 
+used for pretraining the BERT model. We will filter out short lines to ensure that the 
+data has enough context for training.
+"""
+
+keras.utils.get_file(
+    origin="https://s3.amazonaws.com/research.metamind.io/wikitext/wikitext-2-v1.zip",
+    extract=True,
+)
+wiki_dir = os.path.expanduser("~/.keras/datasets/wikitext-2/")
+
+# Load wikitext-103 and filter out short lines.
+wiki_train_ds = (
+    tf.data.TextLineDataset(wiki_dir + "wiki.train.tokens")
+    .filter(lambda x: tf.strings.length(x) > 100)
+    .batch(PRETRAINING_BATCH_SIZE)
+    .cache()
+    .prefetch(tf.data.AUTOTUNE)
+)
+wiki_val_ds = (
+    tf.data.TextLineDataset(wiki_dir + "wiki.valid.token")
+    .filter(lambda x: tf.strings.length(x) > 100)
+    .batch(PRETRAINING_BATCH_SIZE)
+    .cache()
+    .prefetch(tf.data.AUTOTUNE)
+)
+wiki_test_ds = (
+    tf.data.TextLineDataset(wiki_dir + "wiki.test.tokens")
+    .filter(lambda x: tf.strings.length(x) > 100)
+    .batch(PRETRAINING_BATCH_SIZE)
+    .cache()
+    .prefetch(tf.data.AUTOTUNE)
+)
+
+"""
+In the above code, we download the wikitext-2 dataset and extract it. Then, we define 
+three datasets: wiki_train_ds, wiki_val_ds, and wiki_test_ds. These datasets are 
+filtered to remove short lines and are batched for efficient training.
+"""
+
+
+class PrintLR(tf.keras.callbacks.Callback):
+    def on_epoch_end(self, epoch, logs=None):
+        print(
+            f"\nLearning rate for epoch {epoch + 1} is {model_dist.optimizer.lr.numpy()}"
+        )
+
+
+"""
+Let's also make a callback to TensorBoard, this will enable visualization of different metrics
+while we train the model in later part of this tutorial We put all the callbacks together as follows:
+"""
+callbacks = [
+    tf.keras.callbacks.TensorBoard(log_dir="./logs"),
+    PrintLR(),
+]
+
+
+print(tf.config.list_physical_devices("GPU"))
```

Comment:
This will only show a single GPU right on colab right? we might actually want to use this https://www.tensorflow.org/api_docs/python/tf/config/set_logical_device_configuration

You can get the total memory to partition with
https://www.tensorflow.org/api_docs/python/tf/config/experimental/get_memory_info

That will help show a proper two device setup, on a single colab, which I think will be useful to people